File: application/backend-credit-card-enrollment/backend-c#/CreditCard/Enrollment/Projection/EnrollmentList/EnrollmentRepository.cs
using CreditCardEnrollment.Common.Projection;
using MongoDB.Driver;

namespace CreditCardEnrollment.CreditCard.Enrollment.Projection.EnrollmentList;

public class EnrollmentRepository {
    private readonly MongoTransactionalProjectionOperator _mongoOperator;
    private static string _collectionName = "CreditCard_Enrollment_Enrollment";

    public EnrollmentRepository(MongoTransactionalProjectionOperator mongoOperator) {
        _mongoOperator = mongoOperator;
    }

    public void Save(Enrollment enrollment) {
        _mongoOperator.ReplaceOne(
            _collectionName,
            e => e.Id == enrollment.Id,
            enrollment,
            new ReplaceOptions { IsUpsert = true }
        );
    }

    public Enrollment? FindOneById(string id) {
        return _mongoOperator.Find<Enrollment>(
            _collectionName,
            e => e.Id == id
        ).FirstOrDefault();
    }

    public IEnumerable<Enrollment> FindAllByUserId(string userId) {
        return _mongoOperator.Find<Enrollment>(
            _collectionName,
            e => e.UserId == userId
        ).ToEnumerable();
    }
}

File: application/backend-credit-card-enrollment/backend-typescript/src/creditCard/enrollment/projection/enrollmentList/EnrollmentRepository.ts
import {Enrollment} from "./Enrollment";
import {MongoTransactionalProjectionOperator} from "../../../../common/projection/MongoTransactionalProjectionOperator";

export class EnrollmentRepository {
    private readonly collectionName = 'CreditCard_Enrollment_Enrollment';

    constructor(private readonly mongoOperator: MongoTransactionalProjectionOperator) {}

    async save(enrollment: Enrollment): Promise<void> {
        await this.mongoOperator.replaceOne(
            this.collectionName,
            { _id: enrollment._id },
            enrollment,
            { upsert: true }
        );
    }

    async findOneById(_id: string): Promise<Enrollment | null> {
        const results = await this.mongoOperator.find<Enrollment>(this.collectionName, { _id });
        return results.length > 0 ? results[0] : null;
    }

    async findAllByUserId(userId: string): Promise<Enrollment[]> {
        return this.mongoOperator.find<Enrollment>(this.collectionName, { userId });
    }
}